### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by JerzySpendel**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9
```
Comment (grzegorz-wcislo) [inline Dockerfile]: Maybe we can use the slim version instead

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN apt-get update -y
```
Comment (grzegorz-wcislo) [inline Dockerfile]: It's better to use something like:

```
RUN apt-get update -y \
  && apt-get upgrade -y \
  && apt get install -y --no-install-recommends postgresql-client \
  && rm -rf /var/lib/apt/lists/*
```

Using one RUN command results in fewer layers and a smaller image size

The rm at the end removes some not needed stuff that also reduces the image size

the `--no-install-recommends` prevents apt from installing more not needed stuff

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt-get upgrade -y
+RUN apt-get install postgresql-client -y
+
+COPY . /code
```
Comment (grzegorz-wcislo) [inline Dockerfile]: We should also add a dockerignore file to prevent docker from copying not needed stuff from the repo (for example the `.git` folder or the nginx configuration files and whatnot.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+RUN chmod +x /code/wait-for-it.sh
+WORKDIR /code
+
+RUN pip install poetry
+RUN poetry config virtualenvs.create false
+RUN poetry install --no-root
```
Comment (grzegorz-wcislo) [inline Dockerfile]: We should probably also add a user and not run everything as root, I think this is an issue for later though

File: physionet-django/project/admin.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from django.contrib import admin
 from django.db.models import CharField, TextField
 from django.forms import TextInput, Textarea
-from background_task.models_completed import CompletedTask
+#from background_task.models_completed import CompletedTask
```
Comment (grzegorz-wcislo) [inline physionet-django/project/admin.py]: Why did you remove it?

File: wait-for-it.sh [hunk 6]
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env bash
```
Comment (grzegorz-wcislo) [inline wait-for-it.sh]: Very nice addition, i really like this script 😛

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,37 @@
+version: "3.9"
```
Comment (grzegorz-wcislo) [inline docker-compose.yml]: This docker-compose is intended only for local development, right?

I'd recommend adding 2 more volumes (named volumes, not the directory ones), one for media (/code/media) and one for the database.

**Review by grzegorz-wcislo — CHANGES_REQUESTED** [review]
Nice work :)

I request changes because of the Dockerfile and dockerignore stuff. Also please remember to add trailing whitespaces to all the files (github shows a missing one with the red circle at the end of the file)

File: physionet-django/physionet/settings/settings.py [hunk 8]
```diff
@@ -1,45 +1,47 @@
...
 EMAIL_PORT = 25
 EMAIL_HOST_USER = ''
 EMAIL_HOST_PASSWORD = ''
 EMAIL_USE_TLS = False
 
 EMAIL_FROM_DOMAINS = ['physionet.org']
-DEFAULT_FROM_EMAIL = 'PhysioNet Automated System <noreply@physionet.org>'
-CONTACT_EMAIL = 'PhysioNet Contact <contact@physionet.org>'
-SERVER_EMAIL = 'PhysioNet System <root@physionet.org>'
+DEFAULT_FROM_EMAIL = config("DEFAULT_FROM_EMAIL")
```
Comment (rafalcymerys) [inline physionet-django/physionet/settings/settings.py]: It would be great to use old values as default ones, not to break compatibility for physionet's existing deployments

File: physionet-django/physionet/settings/settings.py [hunk 9]
```diff
@@ -1,45 +1,47 @@
...
+        }
     }
 }
 
 # When ready, use the following:
-EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
-EMAIL_HOST = 'localhost'
+EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend' if DEBUG else 'django.core.mail.backends.smtp.EmailBackend'
+EMAIL_HOST = config("EMAIL_HOST")
 EMAIL_PORT = 25
```
Comment (rafalcymerys) [inline physionet-django/physionet/settings/settings.py]: Let's also parameterize SMTP port etc. - if we want to use AWS SES, we'll definitely need a different port + authentication information

File: Dockerfile (hunk 1 above)
Comment (JerzySpendel) [reply Dockerfile]: Some packages need C compiler which is not there in slim version

File: Dockerfile (hunk 4 above)
Comment (JerzySpendel) [reply Dockerfile]: So far I have been assuming that it's fine to do everything as a root in docker containers and I think this is accepted convention to do so? :D

This `--no-root` option here is not what it might look like, it's just as parameter that prevents installing this whole application as a package to current venv. If I didn't pass this `--no-root` option then poetry would install whole `physionet` as a package and then I could do `import physionet` in current venv (which would be the global python interpreter in that case). Poetry is often used as a packaging framework if somebody write a library/framework or something, in these cases it's useful.

File: docker-compose.yml (hunk 7 above)
Comment (JerzySpendel) [reply docker-compose.yml]: There's already volume for /code, so why another for /code/media?

File: wait-for-it.sh (hunk 6 above)
Comment (JerzySpendel) [reply wait-for-it.sh]: :D

File: physionet-django/project/admin.py (hunk 5 above)
Comment (JerzySpendel) [reply physionet-django/project/admin.py]: Ok, uncommented again. I had too new version which didn't have such module `models_completed` and I was wondering why it's not working for me. So I figured out in `django-background-tasks` 1.2.0 there IS `models_completed` module and in 1.2.6 there IS NOT which is a breaking change in a x.y.Z version change, not the best thing to do.

File: Dockerfile (hunk 4 above)
Comment (grzegorz-wcislo) [reply Dockerfile]: There are some security concerns when running as root:

https://security.stackexchange.com/questions/176206/docker-runs-container-processes-as-root-should-i-be-worried
https://americanexpress.io/do-not-run-dockerized-applications-as-root/
